File: test_singles.py
import os
import platform
from copy import copy

import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from scipy import stats
from scipy.optimize import curve_fit


def exp1(x, a, b, c):
    """Compute the exponential function f(x) = a*exp(b*x) + c for input array x."""
    return a * np.exp(b * x) + c


def calibrate_energy(x):
    """Convert energy from samples to keV using known calibration points and an exponential fitting function."""
    px = np.array([0.0, 150, 173, 207])  # samples
    py = np.array([0.0, 334, 511, 1274])  # keV (344 keV compton edge)

    fit1, cov1 = curve_fit(exp1, px, py, p0=(7.0, 0.03, 0.0))
    # xspan = np.linspace(px.min(), px.max())
    # plt.plot(px, py, '.')
    # plt.plot(xspan, exp1(xspan, *fit1))

    label = "{:.3g} * exp({:.3g} * x) + {:.3g}".format(*tuple(fit1))
    return exp1(x, *fit1).clip(max=2000), label


def test_singles(file="my_data_singles.txt"):
    """Tests single events data from a file, calibrates energy, identifies 511 keV candidates, and generates diagnostic
    plots.
    """
    x = np.loadtxt(file)
    t, e, ch = x.transpose()  # time (ps), energy (samples), channel
    print(f"Read {file}, {x.shape[0]:g} events recorded over {(t.max() - t.min()) / 1e12:g}s")

    # calibrate energy
    e0 = copy(e)
    x[:, 1], efit_label = calibrate_energy(e)  # keV
    t, e, ch = x.transpose()  # time (ps), energy (samples), channel
    ecal = copy(e)

    # 511 keV candidates
    x = x[(e > 400) & (e < 600)]  # 400 keV < e_val < 600 keV

    # sort by time
    x = x[np.argsort(x[:, 0])]
    t, e, ch = x.transpose()  # time (ps), energy (keV), channel

    # find pairs
    maxdt = 1000  # (ps)
    dt = np.diff(t)
    a = ch[1:] == 309
    dt[a] = -dt[a]
    dt = dt[(np.abs(dt) < maxdt) & (np.diff(ch) != 0)]

    # Plots
    plots(x, e0, ecal, dt, efit_label)


def plots(x, e0, ecal, dt, efit_label):
    """Generate and save various matplotlib and seaborn plots for given energy and time data arrays."""
    # fig, ax = plt.subplots(2, 3, figsize=(9, 6))
    # ax = ax.ravel()
    # ax[0].hist(e0, np.linspace(0, 300, 300))
    # ax[0].set_xlabel('energy (samples)')
    # ax[1].hist(ecal, np.linspace(0, 2000, 300))
    # ax[1].set_xlabel('energy (keV)')
    # ax[2].hist(dt, np.linspace(-400, 600, 50))
    # ax[2].set_xlabel('t (ps)')
    # fig.tight_layout()
    # # for i in range(3):
    # #     ax[i].grid()
    # fig.savefig('results.png', dpi=200)

    # seaborn correlogram
    # import pandas as pd
    # x = pd.DataFrame(x, columns=['time (ps)', 'E (samples)', 'dt'])
    # sns.pairplot(x, corner=True, diag_kind='hist', kind='scatter', markers='o',
    #              plot_kws=dict(s=3, edgecolor=None, linewidth=1, alpha=0.02),
    #              diag_kws=dict(bins=50))
    # plt.savefig('data_correlogram.png', dpi=200)
    # plt.close()

    # seaborn plots
    fig, ax = plt.subplots(1, 3, figsize=(12, 4))
    ax = ax.ravel()
    sns.distplot(e0, np.linspace(0, 300, 300), kde=False, axlabel="energy (samples)", ax=ax[0])
    sns.distplot(ecal, np.linspace(0, 1500, 300), kde=False, axlabel="energy (keV)", ax=ax[1])
    a = sns.distplot(
        dt,
        np.linspace(-400, 600, 50),
        kde=False,
        fit=stats.norm,
        axlabel="time (ps)",
        label="Normal fit $\mu=${:.1f}, $\sigma=${:.2f}".format(*stats.norm.fit(dt)),
        ax=ax[2],
    )
    a.legend()
    fig.tight_layout()
    fig.savefig("time_resolution.png", dpi=200)

    # seaborn energy calibration jointplot
    import pandas as pd

    x = pd.DataFrame(np.stack((e0, ecal)).transpose(), columns=["energy (samples)", "energy (keV)"])
    a = sns.jointplot(
        data=x,
        x="energy (samples)",
        y="energy (keV)",
        xlim=[0, 300],
        ylim=[0, 1500],
        joint_kws=dict(gridsize=50),
        ratio=2,
        kind="hex",
        marginal_kws=dict(bins=500),
    )
    a.ax_joint.text(10, 1400, efit_label)
    plt.savefig("energy_calibration.png", dpi=200)

    # Display to screen
    if platform.system() == "Darwin":  # MacOS
        os.system("open energy_calibration.png")


if __name__ == "__main__":
    test_singles()

File: testSingles.m
% Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license

function [] = testSingles()
clc; close all
A=fcnloadtextfile('my_data_singles.txt');  X=A.x;
tspan=diff(minmax3(X(:,1)))/1E12;
X = calibrateEnergy(X);  na=size(X,1);

%SELECT 511keV CANDIDATES
e1=400; % (keV)
e2=600; % (keV)
i=X(:,2)>e1 & X(:,2)<e2;  X=X(i,:);  nb=sum(i);

%SET t1=0
X=sortrows(X,1);
X(:,1) = X(:,1)-X(1,1);

%FIND PAIRS
maxdt=1000; %(ps)
dt=diff(X(:,1));
a=X(2:end,3)==309;  dt(a)=-dt(a);
i=find(abs(dt)<maxdt & diff(X(:,3))~=0);  nc=numel(i);
sca; fhistogram(dt(i),50); xlabel(sprintf('dt_{%g-%g keV} (ps)',e1,e2))

fprintf('\n%g singles over %.4f seconds (%.1f Hz)\n%g energy candidate singles (%g-%g keV)\n%g dt candidate pairs (<%g ps dt)\n',na,tspan,na/tspan,nb,e1,e2,nc,maxdt);


function X = calibrateEnergy(X)
%PEAKS ARE AT: 
%511keV = 173
%1274keV = 207
%344keV = 150 (compton edge)
E=X(:,2);
px=[0 150 173 207];
py=[0 334 511 1274];
load Efits.mat

a=Exp1fit(E);  a=min(a,2000);
b=Exp2fit(E);  b=min(b,2000);
c=PCHIPfit(E); c=min(c,2000);

fig(2,3); x=linspace(0,1500,300);
sca; histogram(E,linspace(0,300,300)); xyzlabel('E (samples)')
h=sca; plot(px,py,'k.','Markersize',20,'Display','Calibration Points'); h.XLim=[0 300]; h.YLim=[0 1500];
pa=linspace(0,300,1000); plot(pa,Exp1fit(pa),'Display','Exp1 fit'); plot(pa,Exp2fit(pa),'Display','Exp2 fit'); plot(pa,PCHIPfit(pa),'Display','PCHIP fit'); 
fcnlinewidth(2); xyzlabel('E (samples)','E_{calibrated} (keV)'); legend('Location','Best')
sca; histogram(a,x);  xyzlabel('E_{Exp1 calibrated} (keV)')
sca; histogram(b,x);  xyzlabel('E_{Exp2 calibrated} (keV)')
sca; histogram(c,x);  xyzlabel('E_{PCHIP calibrated} (keV)')
X(:,2)=a;
